### PR TITLE
test inverse of D and plot

### PR DESCRIPTION
**Pull request by EnLAI111**

### CODE REVIEW

File: utils/shared.py [hunk 1]
```diff
@@ -5,8 +5,11 @@ def get_l2norm(A, n_iter=100):
     if isinstance(A, np.ndarray):
         return np.linalg.norm(A, ord=2)
     else:
+        if A.shape[0] < A.shape[1]:
+            A = A.T
+        AtA = A.T @ A
```
Comment (agramfort) [inline utils/shared.py:10]:
```suggestion
        if A.shape[0] < A.shape[1]:
            AtA = lambda x: A.T @ (A @ x)
        else:
            AtA = A.T @ A
            AtA = lambda x: AtA @ x
```

File: utils/shared.py [hunk 2]
```diff
@@ -5,8 +5,11 @@ def get_l2norm(A, n_iter=100):
...
         x = np.random.randn(A.shape[1])
         for _ in range(n_iter):
-            x = A.T @ (A @ x)
+            x = AtA @ x
```
Comment (agramfort) [inline utils/shared.py:13]:
```suggestion
            x = AtA(x)
```

File: utils/shared.py [hunk 3]
```diff
@@ -5,8 +5,11 @@ def get_l2norm(A, n_iter=100):
...
             x /= np.linalg.norm(x)
-        return np.sqrt(np.linalg.norm(A.T @ (A @ x)))
+        return np.sqrt(np.linalg.norm(AtA @ x))
```
Comment (agramfort) [inline utils/shared.py:15]:
```suggestion
        return np.sqrt(np.linalg.norm(AtA(x)))
```

Comment (EnLAI111) [reply utils/shared.py:15]:
```python
In [12]: A = np.random.randn(1000, 5000)

In [13]: %timeit get_l2norm(A)
1.78 s ± 27.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [14]: %timeit get_l2norm_2(A)
1.8 s ± 23.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
The new version is a little slower.

File: tests/test_matrices_operations.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    rng = np.random.RandomState(42)
+    p = rng.randint(2, 100)
+    data = np.array([-np.ones(p), np.ones(p)])
+    diags = np.array([0, 1])
+    D = spdiags(data, diags, p-1, p)
+    D_inv = get_inverse_D(p)
+    np.testing.assert_almost_equal(D @ D_inv, np.identity(p-1))
+
+
+def test_LO_solve_dual_eq():
```
Comment (tomMoral) [inline tests/test_matrices_operations.py:19]: can you add a comment to explain what you are testing here?

Also, I was more into directly testing the code from `DPGD`, by calling it for 3 iterations with a small and dense `A` and then with a linear operator `LinearOperator(shape=A.shape, matvec=lambda x: A@x, ...)`. and checking that the results is the same.

I think you could actually to this for most solvers.

**Review by tomMoral — COMMENTED** [review]
A few more fixes before merging.

File: tests/test_DPGD.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+        if use_acceleration:
+            v_acc[:] = v + (t_old - 1.) / t_new * (v - v_old)
+
+        if isinstance(A, np.ndarray):
+            u = AtA_inv @ (Aty + np.diff(v, append=0, prepend=0))
+        else:
+            u, _ = cg(AtA, Aty + np.diff(v, append=0, prepend=0),
+                      x0=u, tol=tol_cg)
+    return u
```
Comment (tomMoral) [inline tests/test_DPGD.py:58]: Here, you should call  the `run` method of DPGD no? The aim of the test is to check that the code running is correct.

**Review by tomMoral — COMMENTED** [review]
Still a small change left ^^

**Pull request merged**